README.md
Efficient lightweight client-library for easy fluent http/rest requests in Java.

> :bulb: This project has been integrated into [de.galan:commons](https://github.com/galan/commons). For now the project-page acts as landing-page, but will refer to the new location.
> 
> To migrate simply change the dependency to `de.galan:commons` and update the package from `de.galan.flux` to `de.galan.commons.net.flux`.

# Features
* HTTP methods GET, PUT, POST, DELETE, HEAD, TRACE and OPTIONS
* Basic-Authentication support
* Proxy support (with authentication)
* Configurable retries
* Configurable default timeouts and headers

# Integration with Maven
Just add the following dependency (see [releases](https://github.com/galan/commons/releases) for latest version):

    <dependency>
    	<groupId>de.galan</groupId>
    	<artifactId>commons</artifactId>
    	<version>x.y.z</version>
    </dependency>

# Example usage
### Creating Requests
Simplest case

    Response response = Flux.request("http://host/resource").get();
    System.out.println(response.getStreamAsString()); // Work with response, eg. print body
    response.close();

Using AutoCloseable

    try (Response response = Flux.request("http://host/resource").get()) {
        System.out.println(response.getStreamAsString());
    }

### Methods
Calling the HTTP methods executes the http requests and returns a `Response`.

    Flux.request("http://host/resource").get();

The following methods are available:

* `get()`
* `put()`
* `post()`
* `delete()`
* `head()`
* `trace()`
* `options()`
* `method(Method method)`

Additional each method is available as asynchronous call when postfixed with `Async`. Those methods will return a `Future<Response>`:

* `getAsync()`
* `putAsync()`
* `postAsync()`
* `deleteAsync()`
* `headAsync()`
* `traceAsync()`
* `optionsAsync()`
* `methodAsync(Method method)`

    
### Query parameters
You can add parameters by calling the fluent `parameter`methods, adding multiple parameter with the same name is possible.

    Flux.request("http://host/resource").parameter("key", "value").get();
    Flux.request("http://host/resource").parameter("q", "search").paramter("page", "2").get();
    Flux.request("http://host/resource").parameter("cities", "hamburg", "halle").get();

Additional methods for rarely used arguments are also available:

* `paramter(Map<String, String>)`
* `parameterList(Map<String, List<String>>)`.

### Body
You can pass a body by calling the `body(..)` method.

    Flux.request("http://host/resource").body("Hello World").get();
    Flux.request("http://host/resource").body(new byte[]{72, 101, 108, ...}).get();

### HTTP Header
Beside the default header you can set the HTTP header for each request individualy. Methods available in the fluent builder:

* header(String key, String value)
* headers(Map<String, String> headers)

For example:

    Flux.request("http://host/resource").header("X-Myapp", "Demo").header("User-Agent", "Flux").get();
    // code above does the same as below
    Map<String, String> header = new HashMap<>();
    header.put("X-Myapp", "Demo");
    header.put("User-Agent", "Flux");
    Flux.request("http://host/resource").headers(header).get();

### Proxy
Enable a proxy by passing the proxy to a `proxy(..)` method.

    Flux.request("http://host/resource").proxy("1.2.3.4:8080").get();

If the proxy requires authentication use `proxyAuthentication(String username, String password)`.

    Flux.request("http://host/resource").proxy("1.2.3.4:8080").proxyAuthentication("user", "pass").get();

### Timeout
Setting the default connection and read timeout at once (can be overriden using the builder)

    Flux.setDefaultTimeout(2500L); // using 2,5 seconds default
    Flux.setDefaultTimeout("2s500ms"); // using 2,5 seconds default
    Flux.setDefaultTimeout("1m 5s"); // using 1 minutes and 5 seconds default

Setting the timeout using the builder

    Flux.request("http://host/resource").timeout(2500L).get(); // using 2,5 seconds
    Flux.request("http://host/resource").timeout("2s500ms").get(); // using 2,5 seconds
    Flux.request("http://host/resource").timeout("1m 5s").get(); // using 1 minutes and 5 seconds

Setting the connection and read timeout independently

    Flux.setDefaultTimeoutConnection("5s");
    Flux.setDefaultTimeoutRead("1m");
    Flux.request("http://host/resource").timeoutConnection("5s").get();
    Flux.request("http://host/resource").timeoutRead("1m").get();

### Retries
Failed requests can be retried automatically, enable by setting the amount of retries and the time between the retries. If the retries also fail, the last failed response will be returned.

    // repeat 2 times with 10 seconds in between
    Flux.request("http://host/resource").retries(2L, "10s").get(); 

### Redirects
You can follow redirects by calling `followRedirects()`, this is enabled by default. To disable this feature call `unfollowRedirects()`.

## Response
The `Response` returned from the HttpClient contains status information as well as an open connection to retrieve the stream (therefor the required `close()`or `AutoClosable` - benefit: you don't have to retrieve the whole content if you do not require it).

Actions that can be performed on the `Response` object:
* `getStream()` - Returns the raw body `InputStream`
* `getStreamAsBytearray()`- Returns the raw body as Byte array (`byte[]`)
* `getStreamAsString()` - Returns the body as String (UTF-8 is assumed)
* `getStreamAsString(String encoding)` - Returns the body as String (`encoding` is used to decode the String)
* `storeStream(File file)` - Stores the raw body in the given file
* `storeStream(File file, boolean gunzip)` - Stores the raw body in the given file with the option to gzip it
* `getContentEncoding()` - Returns the "Content-Encoding" header
* `getContentType()` - Returns the "Content-Type" header
* `getStatusCode()` - Returns the http status code
* `isFailed()` - Returns true if the status code is not succeded
* `isSucceded()` - Returns true if the status code is between 200 and 299
* `isInformational()` - Returns true if the status code is between 100 and 199
* `isRedirection()` - Returns true if the status code is between 300 and 399
* `isClientError()` - Returns true if the status code is between 400 and 499
* `isServerError()` - Returns true if the status code is between 500 and 599
* `getHeaderField(String name)` - Returns the value for a response header
* `getHeaderFields()` - Returns all response headers as `Map<String, String>`
* `close()` - Closes the response (stream); note that `Response` is `Autoclosable`

# Notes & Hints
* You don't have to use the static `Flux` factory class, if you want to Inject the builder using a DI framework such as Guice, you can simply bind an instance of `FluentHttpClient`.
* Flux is thread-safe
* For testing it is useful to know that you can utilize the `MockCommonHttpClient`. You can pass the `MockCommonHttpClient` with canned Responses. Check the `MockTest` class for an example of the usage. 


changes.txt
final
- Integrated into de.galan:commons:0.12.0

0.7.1
- Bugfix for providing the HttpClient

0.7.0
- releasing to central

0.6
- Provided ready-to-use Mocking Implementations

0.5
- Log4j2 instead of Slf4j
- Asynchronous calls for each http-method

0.4
- added URL to request(..)
- Java 8
- commons updated

0.3
- improved proxy ips and hostnames

0.2
- body(String, Charset) added

0.1
- Initial release


src/main/java/de/galan/flux/Response.java
package de.galan.flux;

import static org.apache.commons.lang3.StringUtils.*;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.nio.charset.Charset;
import java.util.Map;
import java.util.zip.GZIPInputStream;

import org.apache.commons.io.IOUtils;
import org.apache.logging.log4j.Logger;

import com.google.common.collect.ImmutableMap;

import de.galan.commons.logging.Logr;


/**
 * Represents the reponse to a http service.
 *
 * @author daniel
 */
public class Response implements AutoCloseable {

	private static final Logger LOG = Logr.get();

	private final static Map<String, String> EMPTY_HEADERS = ImmutableMap.of();

	private final InputStream dataStream;
	private final int statusCode;
	private final String contentEncoding;
	private final String contentType;
	private final HttpURLConnection connection;
	private final Map<String, String> headerFields;


	public Response(HttpURLConnection connection, InputStream dataStream, int statusCode, String contentEncoding, String contentType, Map<String, String> headerFields) {
		this.connection = connection;
		this.dataStream = dataStream;
		this.statusCode = statusCode;
		this.contentEncoding = contentEncoding;
		this.contentType = contentType;
		this.headerFields = (headerFields != null) ? headerFields : EMPTY_HEADERS;
	}


	public InputStream getStream() {
		return dataStream;
	}


	/** Converts the inputstream to a string in UTF-8. Subsequent the inputstream will be empty/closed. */
	public byte[] getStreamAsBytearray() throws IOException {
		byte[] result = null;
		try {
			result = IOUtils.toByteArray(getStream());
		}
		catch (NullPointerException npex) {
			throw new IOException("Timeout has been forced by CommonHttpClient", npex);
		}
		return result;
	}


	/** Converts the inputstream to a string in UTF-8. Subsequent the inputstream will be empty/closed. */
	public String getStreamAsString() throws IOException {
		return getStreamAsString("UTF-8");
	}


	/**
	 * Converts the inputstream to a string with the given encoding. Subsequent the inputstream will be empty/closed.
	 */
	public String getStreamAsString(Charset charset) throws IOException {
		return getStreamAsString(charset.toString());
	}


	/**
	 * Converts the inputstream to a string with the given encoding. Subsequent the inputstream will be empty/closed.
	 */
	public String getStreamAsString(String encoding) throws IOException {
		String result = null;
		try {
			result = IOUtils.toString(getStream(), encoding);
		}
		catch (NullPointerException npex) {
			throw new IOException("Timeout has been forced by CommonHttpClient", npex);
		}
		return result;
	}


	public void storeStream(File file) throws FileNotFoundException, IOException {
		storeStream(file, false);
	}


	public void storeStream(File file, boolean gunzip) throws FileNotFoundException, IOException {
		InputStream is = getStream();
		if (gunzip) {
			is = new GZIPInputStream(is);
		}
		IOUtils.copy(is, new FileOutputStream(file, false));
	}


	public int getStatusCode() {
		return statusCode;
	}


	public String getContentEncoding() {
		return contentEncoding;
	}


	public String getContentType() {
		return contentType;
	}


	/** Returns true if the http status code is not between 200 and 299 */
	public boolean isFailed() {
		return !isSucceded();
	}


	/** Returns true if the http status code is not between 100 and 199 */
	public boolean isInformational() {
		return getStatusCode() >= 100 && getStatusCode() <= 199;
	}


	/** Returns true if the http status code is between 200 and 299 */
	public boolean isSucceded() {
		return getStatusCode() >= 200 && getStatusCode() <= 299;
	}


	/** Returns true if the http status code is between 300 and 399 */
	public boolean isRedirection() {
		return getStatusCode() >= 300 && getStatusCode() <= 399;
	}


	/** Returns true if the http status code is between 400 and 499 */
	public boolean isClientError() {
		return getStatusCode() >= 400 && getStatusCode() <= 499;
	}


	/** Returns true if the http status code is between 500 and 599 */
	public boolean isServerError() {
		return getStatusCode() >= 500 && getStatusCode() <= 599;
	}


	public String getHeaderField(String name) {
		return isBlank(name) ? null : getHeaderFields().get(name);
	}


	public Map<String, String> getHeaderFields() {
		return headerFields;
	}


	@Override
	public void close() {
		if (getStream() != null) {
			try {
				getStream().close();
			}
			catch (IOException ex) {
				LOG.info("stream could not be closed");
			}
		}
		if (connection != null) {
			connection.disconnect();
		}
	}

}
